pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,273 @@
+from pennylane.templates import broadcast
+from pennylane.templates.utils import (
+    check_shape,
+    check_wires,
+    check_type,
+    get_shape,
+    check_no_variable,
+)
+
+
+@template
+def zz(parameter, wires):
+    """Template for decomposition of ZZ coupling.
+
+    Args:
+        parameter (float): parameter of z rotation
+        wires (list[int]): qubit indices that the template acts on
+    """
+    CNOT(wires=wires)
+    RZ(2 * parameter, wires=wires[1])
+    CNOT(wires=wires)
+
+
+@template
+def IQPEmbedding(features, wires, n_repeats=1, pattern=None):
+    r"""
+    Encodes :math:`n` features into :math:`n` qubits using diagonal gates of an IQP circuit.
+
+    The embedding has been proposed by `Havlicek et al. (2018) <https://arxiv.org/pdf/1804.11326.pdf>`_.
+
+    The basic IQP circuit can be repeated by specifying ``n_repeats``. Repetitions can make the
+    embedding "richer" through interference.
+
+    .. warning::
+
+        ``IQPEmbedding`` calls a circuit that involves non-trivial classical processing of the
+        features. The ``features`` argument is therefore **not differentiable** when using the template, and
+        gradients with respect to the features cannot be computed by PennyLane.
+
+    An IQP circuit is a quantum circuit of a block of Hadamards, followed by a block of gates that are
+    diagonal in the computational basis. Here, the diagonal gates are single-qubit ``RZ`` rotations, applied to each
+    qubit and encoding the :math:`n` features, followed by two-qubit ZZ entanglers,
+    :math:`e^{-i x_i x_j \sigma_z \otimes \sigma_z}`. The entangler applied to wires ``(wires[i], wires[j])``
+    encodes the product of features ``features[i]*features[j]``. The pattern in which the entanglers are
+    applied is either the default, or a custom pattern:
+
+    * If ``pattern`` is not specified, the default pattern will be used, in which the entangling gates connect all
+      pairs of neighbours:
+
+      |
+
+      .. figure:: ../../_static/templates/embeddings/iqp.png
+          :align: center
+          :width: 50%
+          :target: javascript:void(0);
+
+      |
+
+    * Else, ``pattern`` is a list of wire pairs ``[[a, b], [c, d],...]``, applying the entangler
+      on wires ``[a, b]``, ``[c, d]``, etc. For example, ``pattern = [[0, 1], [1, 2]]`` produces
+      the following entangler pattern:
+
+      |
+
+      .. figure:: ../../_static/templates/embeddings/iqp_custom.png
+          :align: center
+          :width: 50%
+          :target: javascript:void(0);
+
+      |
+
+      Since diagonal gates commute, the order of the entanglers does not change the result.
+
+    Args:
+        features (array): array of features to encode
+        wires (Sequence[int] or int): qubit indices that the template acts on
+        n_repeats (int): number of times the basic embedding is repeated
+        pattern (list): specifies the wires and features of the entanglers
```

Comment:
```suggestion
        pattern (Sequence[int]): specifies the wires and features of the entanglers
```

Reply:
`numpy.array` is still a sequence, right? Per @mariaschuld comment above I'm assuming passing that might cause an error later. Would `(List[int])` be a better type for `pattern`?

Reply:
I might be wrong, but I believe `np.ndarray` is an iterable, but not a sequence.

Reply:
Yes, we need at least the wire pairs in this sequence to be a list, so it would be ``Sequence[List[int]]``. there are problems with tuples in core, and there are problems with arrays in plugins. How can one write this out nicely?

Reply:
I can make the internal type more specific with ``pattern (list[int])``...?

Reply:
The user should not use anything else if possible!